MAINT: stats: tweak the formula for ncx2.pdf

ev-br opened the pull request:
Simplify the formula for `ncx2.pdf`. In master, it uses hyp0f1 which overflows at large values of the argument, (see https://github.com/scipy/scipy/issues/1577), and ncx2 has to use nan_to_num.

This PR rewrites the pdf in terms of the modified Bessel function, and nan_to_num is no longer needed.

WarrenWeckesser commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -612,13 +611,18 @@ def _parse_args_stats(self, %(shape_arg_str)s %(locscale_in)s, moments='mv'):
+    # at large values of xs. See also `rice.pdf`.
+    df2 = df/2.0 - 1.0
+    xs, ns = np.sqrt(x), np.sqrt(nc)
+    res = xlogy(df2/2.0, x/nc)- 0.5*(xs - ns)**2 + np.log(ive(df2, xs*ns) / 2.0)
```
Unmatched whitespace around `-`.

WarrenWeckesser commented:
Nice, thanks.  I'll take a closer look later today.

From a first glance at the test, it looks like it should only be necessary to filter the `RuntimeWarning` that occurs when the `logpdf` method is called with 0s.  The `pdf` method should not generate any warnings, so that part of the test should be moved out of the `with` context.

Does anyone know of a concrete example that's not so far out in the tail where this version gives a better numerical result?  I'll see if I can find one.

WarrenWeckesser commented:
I just noticed that the test makes the same call to `pdf`, once outside the `with` context and once inside.  The one inside the `with` context should be removed.

WarrenWeckesser commented:
Here's an example where the new version is significantly *less* precise than the old code.  The "exact" value is computed using mpmath (with `mpmath.mp.prec = 600`):

"Exact":

```
In [677]: x
Out[677]: mpf('1600.0')

In [678]: k
Out[678]: mpf('100.0')

In [679]: lam
Out[679]: mpf('1.0')

In [680]: float(mpmath.exp(-(x + lam)/2)*mpmath.power(x / lam, (k/2 - 1)/2)*mpmath.besseli(k/2 - 1, mpmath.sqrt(lam*x))/2)
Out[680]: 5.7453950222319135e-266

In [681]: lam = mpmath.mpf('2.0')

In [682]: float(mpmath.exp(-(x + lam)/2)*mpmath.power(x / lam, (k/2 - 1)/2)*mpmath.besseli(k/2 - 1, mpmath.sqrt(lam*x))/2)
Out[682]: 2.801461420251783e-263
```

Old version:
```
In [684]: ncx2.pdf(1600, 100, 1)
Out[684]: 5.7453950222320189e-266

In [685]: ncx2.pdf(1600, 100, 2)
Out[685]: 2.8014614202516076e-263
```

This pull request:
```
In [124]: ncx2.pdf(1600, 100, 1)
Out[124]: 0.0

In [125]: ncx2.pdf(1600, 100, 2)
Out[125]: 2.7774321131400449e-263
```

Admittedly, these values are out near (or even beyond) the edge of usefulness, so perhaps it is not an issue.  I haven't looked into the cause--perhaps a limitation of `ive`?

For many more informal tests, the new code appears to be as accurate as the old version.

ev-br commented:
Thanks Warren!

This is a nice lesson for me --- simply moving the computation to the logspace seems to restore the accuracy. In the force-pushed version which does just that, I get

```
In [2]: ncx2.pdf(1600, 100, 1)
Out[2]: 5.7453950222313661e-266

In [3]: ncx2.pdf(1600, 100, 2)
Out[3]: 2.8014614202512888e-263
```

While not directly relevant to this PR, I should mention that the main issue with ncx2 accuracy is likely in the cdf implementation (`chndtr`). 
https://github.com/scipy/scipy/blob/master/scipy/stats/tests/test_discrete_basic.py#L110

WarrenWeckesser commented:
The problem with this version is that it generates spurious warnings for values far out in the tail, where the PDF function returns 0:

```
In [2]: ncx2.pdf(1, 340, 2)
/Users/warren/local_scipy/lib/python2.7/site-packages/scipy/stats/_distn_infrastructure.py:620: RuntimeWarning: divide by zero encountered in log
  res += np.log(ive(df2, xs*ns) / 2.0)
Out[2]: 0.0
```
That warning is the result of an implementation detail, so it should never be exposed to a user.

ev-br commented:
This can possibly use _lazywhere. Adding a special function for just log(Iv) looks overkill.
[shrug] 
I personally dislike nan_to_num or filtering out warnings in the library code. 

FTR, what is now in master is a bit weird: it uses hyp0f1, which is essentially a modified Bessel function times gamma, and divides it by a gamma function (in the logspace). 
At large values of parameters, gamma overflows, and so is hyp0f1, so it's wrapped into a `nan_to_num`. 
Shrug.

WarrenWeckesser commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -612,15 +611,18 @@ def _parse_args_stats(self, %(shape_arg_str)s %(locscale_in)s, moments='mv'):
 def _ncx2_pdf(x, df, nc):
     return np.exp(_ncx2_log_pdf(x, df, nc))
 
-
```
Don't remove this blank line.  See the first sentence in the "Blank Lines" section of PEP 8: https://www.python.org/dev/peps/pep-0008/#blank-lines

WarrenWeckesser commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -612,15 +611,18 @@ def _parse_args_stats(self, %(shape_arg_str)s %(locscale_in)s, moments='mv'):
+    xs, ns = np.sqrt(x), np.sqrt(nc)
+    res = xlogy(df2/2.0, x/nc) - 0.5*(xs - ns)**2
+    res += np.log(ive(df2, xs*ns) / 2.0)
+    return res
```
Add another blank line after the function (see my comment below).

WarrenWeckesser commented:
> I personally dislike nan_to_num or filtering out warnings in the library code. 

I agree.

It occurs to me that with the latest version of this PR, you haven't touched the `pdf` function, which is the function that generates the spurious warnings.   The code in master has the same problem.   You haven't broken anything, and this change is an improvement, so I'm OK if you want to put off fixing the warning to another PR (but create an issue for it if it won't get fixed soon).  I wouldn't be surprised if there are other distributions with the same problem.

ev-br commented:
Fixed the whitespace. 

> It occurs to me that with the latest version of this PR, you haven't touched the pdf function,

Indeed. I've reverted back to computing the pdf in logspace after you've shown the issue with the accuracy far out in the tails. The current version only changes logpdf.

WarrenWeckesser commented:
OK, merged.

WarrenWeckesser commented:
Issue for the spurious warnings: https://github.com/scipy/scipy/issues/5202
